arch/xtensa/src/esp32/esp32_rmt.c

```diff
@@ -0,0 +1,315 @@
+  modifyreg32(DPORT_PERIP_RST_EN_REG, DPORT_RMT_RST, 0);
+  putreg32(0xFFFFFFFF, RMT_INT_CLR_REG); /* Clear any spurious IRQ Flag */
+
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: esp32rmt_setup
+ *
+ * Description:
+ *   Configure the RMT. This method is called the first time that the RMT
+ *   device is opened.  This will occur when the port is first opened.
+ *   This setup includes configuring and attaching RMT interrupts.
+ *
+ * Input Parameters:
+ *   dev - An instance of the "upper half" RMT driver state structure.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int esp32rmt_setup(struct rmt_dev_s *dev)
+{
+  irqstate_t flags;
+  int ret = OK;
+
+  flags = enter_critical_section();
+
+  if (dev->cpuint != -ENOMEM)
+    {
+      /* Disable the provided CPU Interrupt to configure it. */
+      up_disable_irq(dev->cpuint);
+    }
+
+  dev->cpu = up_cpu_index();
+  dev->cpuint = esp32_setup_irq(dev->cpu, dev->periph,
+                                2, ESP32_CPUINT_LEVEL);
+  if (dev->cpuint < 0)
+    {
+      /* Failed to allocate a CPU interrupt of this type. */
+  
+      ret = dev->cpuint;
+      leave_critical_section(flags);
+  
+      return ret;
+    }
+  
+  ret = irq_attach(dev->irq, esp32rmt_interrupt, dev);
+  
+  if (ret != OK)
+    {
+      /* Failed to attach IRQ, so CPU interrupt must be freed. */
+  
+      esp32_teardown_irq(dev->cpu, dev->periph, dev->cpuint);
+      dev->cpuint = -ENOMEM;
+      leave_critical_section(flags);
+  
+      return ret;
+    }
+
+  /* Enable the CPU interrupt that is linked to the RMT device. */
+  up_enable_irq(dev->cpuint);
+
+  leave_critical_section(flags);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32rmt_interrupt
+ *
+ * Description:
+ *   RMT TX interrupt handler
+ *
+ * Input Parameters:
+ *   irq - The IRQ number of the interrupt.
+ *   context - The register state save array at the time of the interrupt.
+ *   arg - The pointer to driver structure.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno on failure
+ *
+ ****************************************************************************/
+IRAM_ATTR static int esp32rmt_interrupt(int irq, void *context, void *arg)
+{
+  struct rmt_dev_s *dev = (struct rmt_dev_s *)arg;
+  uint32_t regval = getreg32(RMT_INT_ST_REG);
+  uint32_t total_pixels;
+  uint32_t current_pixel;
+  uint32_t dst_mem;
+  uint32_t used_words;
+  uint8_t *pixel_ptr;
+  
+  if (regval & RMT_CHn_TX_END_INT_ST(dev->channel))
+    {
+      putreg32(RMT_CHn_TX_END_INT_CLR(dev->channel), RMT_INT_CLR_REG);
+      regval = getreg32(RMT_INT_ENA_REG);
+      regval &= ~(RMT_CHn_TX_END_INT_ENA(dev->channel));
+      putreg32(regval, RMT_INT_ENA_REG);
```

Comment:
```suggestion

      modifyreg32(RMT_INT_ENA_REG, RMT_CHn_TX_END_INT_ENA(dev->channel), 0);
```
`modifyreg32` can be used for achieving cleaner code.

Reply:
Yes, I agree.
I've tryed to avoid it inside de interrupt routine since in those cases there is no need for atomic setting/clearing of the register's value. Also, it has some spinlock inside it. Isn't it a bit slower then getreg32+putreg32?
I can change it if you insist, no hard feelings.

Reply:
> Isn't it a bit slower then getreg32+putreg32?

Yes, but the performance penalty is negligible.
When not in SMP mode, it is just a write to (read from) a special register.
Under SMP mode it indeed spins in a lock, but I wouldn't worry about this micro-optimization if not really required.